[dynamo] fix dynamo + DTensor to work with 2d

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #108329

pair debugged with @wconstab and we found some issue in both dynamo and
the TP's fsdp extension side. This PR fixes the dynamo + DTensor integration
so that the current graph break FSDP can work with tensor parallel by moving
the torch.compile after FSDP wrapping.

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov

wanchaol committed:
[dynamo] fix dynamo + DTensor to work with 2d

This PR fixes the dynamo + DTensor integration so that the current graph
break FSDP can work with tensor parallel by moving the torch.compile
after FSDP wrapping

[ghstack-poisoned]

wanchaol committed:
Update on "[dynamo] fix dynamo + DTensor to work with 2d"


pair debugged with wconstab and we found some issue in both dynamo and
the TP's fsdp extension side. This PR fixes the dynamo + DTensor integration
so that the current graph break FSDP can work with tensor parallel by moving
the torch.compile after FSDP wrapping.

[ghstack-poisoned]

wanchaol committed:
Update on "[dynamo] fix dynamo + DTensor to work with 2d"


pair debugged with wconstab and we found some issue in both dynamo and
the TP's fsdp extension side. This PR fixes the dynamo + DTensor integration
so that the current graph break FSDP can work with tensor parallel by moving
the torch.compile after FSDP wrapping.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

[ghstack-poisoned]

wconstab commented on test/distributed/_tensor/test_dtensor_compile.py:
```diff
@@ -161,22 +161,18 @@ def test_2d_fsdp_tp_compile(self):
         )
 
+        # TODO: once aot autograd support is ready we can just use default backend
+        compiled_2d = torch.compile(fsdp_2d, backend="eager")
```
nit: we might also want to test inductor backend, as it does break sometimes more than eager, due to customized logic for some parts of aot-autograd

wanchaol replied:
aot autograd is currently broken until @bdhirsh aot dispatch subclass PR landed https://github.com/pytorch/pytorch/pull/104483, will switch this test to test both aot_eager or default backend once it's there :)

wconstab commented on torch/_dynamo/variables/torch.py:
```diff
@@ -572,6 +572,7 @@ def get_state_from_generator():
         elif is_from_local(self.value):
             # rewrite non-primitive args/kwargs to be included in the on-the-fly prim function
             # and rewrite args to have only proxyable args, then insert call_function
+            # TODO: support cases where device_mesh + placements specified as kwargs
```
mby good to assert kwargs is empty (or that it just contains the 1 bool you expect, etc)

wanchaol replied:
Actually it's not quite hard to directly support kwargs so I added that directly

wconstab reviewed (approved):
lgtm, but definitely add the assert for kwargs, might even be easy enough to just support kwargs too

wanchaol committed:
Update on "[dynamo] fix dynamo + DTensor to work with 2d"


pair debugged with wconstab and we found some issue in both dynamo and
the TP's fsdp extension side. This PR fixes the dynamo + DTensor integration
so that the current graph break FSDP can work with tensor parallel by moving
the torch.compile after FSDP wrapping.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

[ghstack-poisoned]

bdhirsh commented:
Nice!
